### PR TITLE
Basic working implementation of forgiving join with output provenance structure, and rudiments of pipeline system

### PR DESCRIPTION
**Pull request by amb26**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,11 +7,139 @@ maintaining rather than effacing provenance.
...
+demanded,
+2. The provenance of each final object in the merged structure is available in a "provenance map" which is isomorphic
+to the merged value, recording which layer each resulting leaf value was drawn from
+
+The primary API of the mat consists of its creator function `fluid.tangledMat`, and methods `addLayer`, `getRoot`,
+`readMember`, `getProvenance`, `evaluateFully` and `setWritableLayer` - see JSDocs. An example session interacting with the
+mat mirroring the example drawing:
+
+````javascript
+var mat = fluid.tangledMat([{
```
Comment (cindyli) [inline README.md:49]: What if different layers contain the same element, for example, key `a` exists in "Layer2" or/and "Layer3"? In this case, is the plan to ask for the user intervention or allow defining the precedence at adding layers? Is this use case not yet considered in this version?

File: README.md [hunk 2]
```diff
@@ -7,11 +7,139 @@ maintaining rather than effacing provenance.
...
+    {
+        value: <Array of CSV row values, as loaded from the dataset, with each row as a hash>
+        provenance: <Isomorphic to value, with a provenance string for each data value - as per tangled mat's provenance> 
+        provenanceMap: <A map of provenance strings to records resolving the provenance - either a dataset record or another pipeline record>
+    }
+````
+
+Pipeline elements available include
+
+    fluid.forgivingJoin - executes an inner or outer join given two CSV structures
```
Comment (cindyli) [inline README.md:109]: This pipeline seems supporting inner or outer right join, inner or outer left join. Can you explain what they are or add a link to the explanations? Thanks.

File: README.md [hunk 3]
```diff
@@ -7,11 +7,139 @@ maintaining rather than effacing provenance.
...
+        repository: <git repository URL, e.g. "https://github.com/inclusive-design/covid-assessment-centres">
+        path: <String path within the repository, e.g. "WeCount/assessment_centre_data_collection_2020_09_02.csv">
+    },
+
+    pipeline: Free hash of pipeline elements with elements
+        type: <String - global function name>
+        <other type-dependent properties>
+````
+
+Pipeline elements accept and return records as the following triple:
```
[the remark this answers is not in the document]
Comment (cindyli) [reply README.md:97]: I see. These are result files created by the pipeline, not descriptions for elements in the `output` block of the config file.

File: src/dataPipeline.js [hunk 4]
```diff
@@ -113,19 +132,21 @@ fluid.data.executePipeline = function (job) {
         if (fluid.hasGrade(grade, "fluid.simpleInputPipe")) {
             var prevOutput = prevKey ? pipeOutputs[prevKey] : {};
             input = onePipe.input === "_" ? prevOutput : pipeOutputs[onePipe.input];
```
Comment (cindyli) [inline src/dataPipeline.js:134]: The usage of `input: "_"` in the config file worths a good explanation. I was very curious when seeing it in the config file until I read this line.

File: README.md (hunk 1 above)
Comment (amb26) [reply README.md:49]: Values in higher layers always override those in lower ones, just as with $.extend(true ....)  - I've added a test case to clarify this

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (amb26) [reply README.md:97]: Thanks, this was unclear - I've reorganised this section of the README and added a typedef in the JSDocs

File: README.md [hunk 5]
```diff
@@ -7,11 +7,139 @@ maintaining rather than effacing provenance.
...
+//     }
+// }
+````
+
+### The data pipeline
+
+The pipeline is configured by a JSON5 structure with top-level elements
+
+````text
+    type {String} The name of the overall pipeline (will eventually become a grade name of some sort)
```
[the remark this answers is not in the document]
Comment (amb26) [reply README.md:84]: Added

File: README.md (hunk 2 above)
Comment (amb26) [reply README.md:109]: Added

File: src/dataPipeline.js (hunk 4 above)
Comment (amb26) [reply src/dataPipeline.js:134]: This is explained in the README under fluid.simpleInputPipe at https://github.com/inclusive-design/forgiving-data/blob/a24d0899757531d9ddb2fc43192d6a4267305b2c/README.md#fluidsimpleinputpipe

File: src/forgivingJoin.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            var record = fluid.data.intersect(leftHash, rightHash);
+            record.leftKey = leftKey;
+            record.rightKey = rightKey;
+            intersects.push(record);
+        });
+    });
+    intersects.sort(function (inta, intb) {
+        return intb.count - inta.count;
+    });
+    var best = intersects[0];
```
Comment (cindyli) [inline src/forgivingJoin.js:119]: Using the best matched column assumes only one single column serves as the primary key to join 2 datasets. It doesn't cover the use case of the composite primary key where more than one columns are used to form the primary key to identify matched records in 2 datasets.

Comment (amb26) [reply src/forgivingJoin.js:119]: Thanks for the suggestion - I've updated the README to explain the algorithm used and written up the improvement as https://issues.fluidproject.org/browse/DATA-1

**Pull request merged**